server/src/main/java/com/cloud/vm/UserVmManagerImpl.java

```diff
@@ -2966,9 +2966,20 @@ public UserVm destroyVm(DestroyVMCmd cmd) throws ResourceUnavailableException, C
         checkForUnattachedVolumes(vmId, volumesToBeDeleted);
         validateVolumes(volumesToBeDeleted);
 
+        List<VolumeVO> allVolumes = null;
+        if (vm.getHypervisorType() == HypervisorType.VMware) {
+            allVolumes = _volsDao.findByInstance(vm.getId());
+            allVolumes.removeIf(vol -> vol.getVolumeType() == Volume.Type.ROOT);
+        }else {
+            allVolumes = volumesToBeDeleted;
+        }
+        for (VolumeVO volume : allVolumes) {
+             _accountMgr.checkAccess(ctx.getCallingAccount(), null, true, volume);
+        }
+
```

Comment:
We could probably use the following to check access for all volumes to be deleted?
```suggestion
       final ControlledEntity[] volumesToDelete = volumesToBeDeleted.toArray(new ControlledEntity[0]);
        _accountMgr.checkAccess(ctx.getCallingAccount(), null, true, volumesToDelete);

```